File: greatest-common-divisor-of-strings.rb
# @param {String} str1
# @param {String} str2
# @return {String}
def gcd_of_strings(str1, str2)
  longest = str1.length > str2.length ? str1 : str2  
  count = 0
  ans = ''
  
  while count < longest.length
    str = longest[0..count]
    gcd1 = ''
    gcd2 = ''
    gcd1 += str while (gcd1.length < str1.length)
    gcd2 += str while (gcd2.length < str2.length)

    if gcd1 == str1 && gcd2 == str2
      ans = str
    end
    count+=1
  end
    
  ans
end

# Runtime: O(n)
# Space: O(n+m)

File: greatest-common-divisor-of-strings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */

// O(m(n+m))
// S(n+m)

const gcdOfStrings = (str1, str2) => {
  let ans = '';
  for (let i=1; i<=str2.length; i++) {
    let subset = str2.slice(0,i);
    let str1c = subset;
    let str2c = subset;
    while (str1c.length < str1.length) str1c += subset;  
    while (str2c.length < str2.length) str2c += subset;  
    if (str1c === str1 && str2c === str2) ans = subset;      
  }

  return ans;
};

